### PR TITLE
Api 1407w3

### PR DESCRIPTION
**Pull request by willy-ahva**
**Description (for Contributor and Core Developer)**

Add the test: test_removing_category_from_variant_product_produces_an_event
EE PR: https://github.com/akeneo/pim-enterprise-dev/pull/10457

**Definition Of Done (for Core Developer only)**

| Q                                 | A
| --------------------------------- | ---
| Added Specs                       | Todo
| Added legacy Behats               | Todo
| Added acceptance tests            | Todo
| Added integration tests           | Todo
| Changelog updated                 | Todo
| Review and 2 GTM                  | Todo
| Micro Demo to the PO (Story only) | Todo
| Migration script                  | -
| Tech Doc                          | -

`Todo`: Pending / Work in progress
`OK`: Done / Validated
`-`: Not needed

### CODE REVIEW

File: tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php [hunk 1]
```diff
@@ -37,6 +39,50 @@ protected function authenticate(UserInterface $user): void
...
+        $this->get('akeneo_elasticsearch.client.product_and_product_model')->refreshIndex();
+        $this->get('pim_catalog.validator.unique_value_set')->reset();
+
+        return $product;
+    }
+
+    /**
+     * Each time we create a product model, a batch job is pushed into the queue to calculate the
+     * completeness of its descendants.
+     */
```
Comment (pchasle) [inline tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php:66]:
```suggestion
```
It's interesting to know that, but I would not put this in a comment here as it's a completely decoupled side effect.

File: tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php [hunk 2]
```diff
@@ -37,6 +39,50 @@ protected function authenticate(UserInterface $user): void
         $this->client->getCookieJar()->set($cookie);
     }
 
+    protected function createVariantProduct(string $identifier, array $data = []) : ProductInterface
```
Comment (tomglvng) [inline tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php:42]: I guess, you can create a simple product with this function, correct ? So, you can name it `createProduct`. Or maybe, if you want to keep a `createVariantProduct`, you should make some data mandatory, like `parent`. What do you think ?

File: tests/back/Pim/Enrichment/EndToEnd/Product/VariantProduct/InternalApi/UpdateVariantProductEndToEnd.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+                        'locale' => null,
+                        'scope' => null,
+                        'data' => false,
+                    ],
+                ],
+            ],
+        ]);
+        $normalizedProduct = $this->getProductFromInternalApi((string) $product->getId());
+        $this->clearMessengerTransport();
+        $normalizedProduct['categories'] = array_filter($normalizedProduct['categories'], fn($value) => 'categoryB' !== $value);
```
Comment (pchasle) [inline tests/back/Pim/Enrichment/EndToEnd/Product/VariantProduct/InternalApi/UpdateVariantProductEndToEnd.php:36]: That's a fancy way of doing `$normalizedProduct['categories'] = ['master'];`

**Review by pchasle — APPROVED** [review]

File: tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php (hunk 2 above)
Comment (pchasle) [reply tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php:42]: +1

Comment (pchasle) [reply tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php:42]: What about making this into some container aware traits, so we could reuse them in other TestCase easily ?

Comment (willy-ahva) [reply tests/back/Pim/Enrichment/EndToEnd/InternalApiTestCase.php:42]: @tomglvng I agree.
@pchasle I agree about the trait but it should not be done in my PR as it needs some refactoring on a lot of other classes.

**Review by tseho — APPROVED** [review]

**Pull request merged**